store.go

```diff
@@ -2479,8 +2493,28 @@ func (s *store) DeleteLayer(id string) error {
 func (s *store) DeleteImage(id string, commit bool) (layers []string, err error) {
 	layersToRemove := []string{}
 	if err := s.writeToAllStores(func(rlstore rwLayerStore) error {
-		if s.imageStore.Exists(id) {
-			image, err := s.imageStore.Get(id)
+		// Perform delete image on primary imageStore but if image
+		// is not found on primary store then try delete on any additional
+		// write-able image store.
+		imageStores := []rwImageStore{s.imageStore}
+		imageStores = append(imageStores, s.rwImageStores...)
+		imageFound := false
+		primaryImageStore := true
+		for _, imageStore := range imageStores {
+			if !imageStore.Exists(id) {
+				primaryImageStore = false
+				continue
+			}
+			imageFound = true
+			if !primaryImageStore {
```

Comment:
Shouldn’t this just be `if imageStore != s.imageStore`?

- That’s _much_ much more explicit about what we need to check for. The locking logic _really_ doesn’t care about which store was used first, or whether the image exists.
- It’s much shorter to implement
- This one is actually incorrect: If an image exists in all stores, `primaryImageStore` is never set to `false`, and the other stores incorrectly be used without locking.